src/lib/auth/index.js

```diff
@@ -0,0 +1,122 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import {
+	AUTH_TYPE_WPCOM_OAUTH,
+	AUTH_TYPE_WPCOM_OAUTH_BY_TOKEN,
+	AUTH_TYPE_WPCOM_PROXY_IFRAME,
+} from './strategies';
+import WPcomOAuth from './strategies/wpcom/oauth';
+import WPcomProxyIframe from './strategies/wpcom/proxy-iframe';
+
+/**
+ * Module variables
+ */
+let strategy;
+let auth;
+const debug = debugFactory( 'happychat-client:auth' );
+
+/**
+ * Initialize authentication library (authenticator). Based on the authentication type a strategy
+ * will be prepared.
+ *
+ * The package acts like a singleton. Both the strategy and the authentication options are saved in
+ * the package's object which allows us to only initialize the library once. Meaning that after we
+ * initialize the library it will be shared throughout the entire codebase without the need of
+ * re-initialization.
+ *
+ * Currently implemented strategies are:
+ *  - 'wpcom-oauth'          regular wpcom oAuth process using `wpcom-oauth-cors` package
+ *  - 'wpcom-oauth-by-token' regular wpcom oAuth based on given token using `wpcom-xhr-request`
+ *  - 'wpcom-proxy-iframe'   proxy iframe wpcom authentication using Calypso's wpcom object
+ *
+ * @param {Object} authentication information about the authentication type and its options
+ * @param {string} authentication.type authentication type (one of the implemented strategies)
+ * @param {Object} authentication.options set of options sent to auth implementations
+ * @param {string} authentication.options.token oauth accessToken used by `wpcom-oauth-by-token`
+ * @param {Object} authentication.options.proxy wpcom proxy object used by `wpcom-proxy-iframe`
```

Comment:
Might be nice to add `@throws` statement here as well.

Reply:
👍